### PR TITLE
CLC update strategy

### PR DESCRIPTION
**Pull request by aermakov-zalando**
MVP for Cluster Lifecycle Controller integration. If update strategy is set to `clc`, only update the launch configuration & launch templates, mark the outdated nodes with `decommission-pending` and wait for CLC to remove them.

Note that CLM also does the draining when decommissioning node pools or scaling down; this PR only touches the node update code. Once we're comfortable with using CLC, I'd rather drop the old code completely and refactor the whole `node_pool_manager`/`update_strategy` mess.

### CODE REVIEW

File: provisioner/clusterpy.go [hunk 1]
```diff
@@ -621,47 +622,50 @@ func (p *clusterpyProvisioner) prepareProvision(logger *log.Entry, cluster *api.
...
+		{"drain_poll_interval", func(v time.Duration) { drainConfig.PollInterval = v }},
+	} {
+		if value, ok := cluster.ConfigItems[setting.key]; ok {
+			parsed, err := time.ParseDuration(value)
+			if err != nil {
+				return nil, nil, nil, fmt.Errorf("invalid value for %s: %v", setting.key, err)
+			}
+			setting.fn(parsed)
+		}
+	}
```
Comment (mikkeloscar) [inline provisioner/clusterpy.go]: I'm not sure this change makes it easier to read the code?

File: pkg/updatestrategy/clc_update.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+	}
+}
+
+func (c *CLCUpdateStrategy) Update(ctx context.Context, nodePoolDesc *api.NodePool) error {
+	c.logger.Infof("Initializing update of node pool '%s'", nodePoolDesc.Name)
+
+	err := c.doUpdate(ctx, nodePoolDesc)
+	if err != nil {
+		if ctx.Err() != nil {
+			err := c.rollbackUpdate(nodePoolDesc)
```
Comment (mikkeloscar) [inline pkg/updatestrategy/clc_update.go]: Is the intention that the CLC will handle this in the future? I understand why it's done this way, but it seems like the CLC should do this by itself at some point or is that not the intention?

File: provisioner/clusterpy.go (hunk 1 above)
Comment (aermakov-zalando) [reply provisioner/clusterpy.go]: The change makes it actually handle the errors, previously it ignored those completely. :(

File: pkg/updatestrategy/clc_update.go (hunk 2 above)
Comment (aermakov-zalando) [reply pkg/updatestrategy/clc_update.go]: I don't really have any plans for this, we could move cluster-update blocking to the clusters as well I guess?

File: pkg/updatestrategy/clc_update_test.go [hunk 3]
```diff
@@ -0,0 +1,230 @@
+package updatestrategy
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+
```
Comment (szuecs) [inline pkg/updatestrategy/clc_update_test.go]: please clean empty lines

Comment (aermakov-zalando) [reply pkg/updatestrategy/clc_update_test.go]: Fixed.

**Comment by mikkeloscar** [comment]
:+1:

**Comment by aermakov-zalando** [comment]
👍

**Comment by mikkeloscar** [comment]
:+1:

**Pull request merged**